### PR TITLE
(Update) IRC AnnounceBot

### PR DESCRIPTION
**Pull request by Obi-Wana**
As discussed, here some changes that fixed the IRC bot for me when running `php artisan irc:message ...`.

Debugging this was kinda hard and may be very individual. In my case, adding the EOL to the commands and a bit of `sleep` fixed it and IRC messages were posted to an inspircd test server.

Please note that this is a draft and needs testing and maybe some code cleanups.

### CODE REVIEW

**Comment by costaht** [comment]
Hi @Obi-Wana, thank you for your interest in getting IRC Announce working. I'm also interested in this feature and if you don't mind I'd like to help you by validating the changes you've made in the code.
I replaced my `IRCAnnounceBot.php` with [yours](https://raw.githubusercontent.com/Obi-Wana/UNIT3D-Community-Edition/fork/Fix-IRC-announce/app/Bots/IRCAnnounceBot.php)  but I'm getting some errors as you can see in the logs:
https://pastebin.com/NqBZW37W

**Comment by Obi-Wana** [comment]
> Hi @Obi-Wana, thank you for your interest in getting IRC Announce working. I'm also interested in this feature and if you don't mind I'd like to help you by validating the changes you've made in the code. I replaced my `IRCAnnounceBot.php` with [yours](https://raw.githubusercontent.com/Obi-Wana/UNIT3D-Community-Edition/fork/Fix-IRC-announce/app/Bots/IRCAnnounceBot.php) but I'm getting some errors as you can see in the logs: https://pastebin.com/NqBZW37W


Does that line exist in your file?
https://github.com/HDInnovations/UNIT3D-Community-Edition/pull/2582/files#diff-2d843516199b4212785d5f926d992b8fbe7c8036bf2ea632c90fed49ac47bceeR53

Your error message says, that this variable is undefined.

**Comment by costaht** [comment]
> Your error message says, that this variable is undefined.

yes, it does. I just `wget link -O IRCAnnounceBot.php` and ajusted the file's permission.

Screenshot from 2023-03-21 15-03-31

I also think that error is odd since the variable is declared at the top of the file. Isn't that because the variable is inside of a function?

**Comment by HDVinnie** [comment]
@costaht try latest code in this PR

**Comment by costaht** [comment]
Thank you guys for helping me with that. **It works!!!**

@HDVinnie here are some insights:

On line 92 you need to add a colon before the variable `string` because all messages have to [start with](https://modern.ircdocs.horse/#message-format) `:`. 
 Ex: `$this->send_data(sprintf('PRIVMSG %s :%s %s', $channel, $string, $this->eol));`
Without it only the first word of a sentence will be sent.

To whomever is testing this. Pass the parameters between double quotes like below:
`php artisan irc:message "#announce" "This is a test 1 2 3"`

About `if ($ex[1] === "366")`. This tells the bot to wait for `RPL_ENDOFNAMES (366)` which is the [list of users](https://modern.ircdocs.horse/#rplendofnames-366) of a channel, but you will only get that code IF you join a channel. 

So basically if the user set  `irc-bot.joinchannel` to `false`, they will never get that code back from the server and the message will only be sent after 60 seconds (some sort of timeout).

Since my bot won't join the announce channel I changed mine to `376`, which is [RPL_ENDOFMOTD](https://modern.ircdocs.horse/#rplendofmotd-376), the response code for the server's MOTD. Now each message takes two seconds to be delivered (sleep(2)).

To finish, I'm gonna format the message so users can make use of autobrr which will make announces much faster and alleviate the number of users hammering the server consulting the RSS.

Thank you very much again, and please let me know if you need me to make any other tests.

**Comment by costaht** [comment]
Also we need to import password:

`$this->password = config('irc-bot.password');
`
And change this 
`$this->send_data(sprintf('NICKSERV IDENTIFY %s', $this->nickservpass));
`
to this:
`$this->send_data(sprintf('NICKSERV IDENTIFY %s', $this->password));`

**Comment by costaht** [comment]
@d0rq I don't wanna spam here. Reach me out on Discord tijuco#5335

**Comment by Obi-Wana** [comment]
> @costaht I did all the things you said to do, however the bot connects for those 2 secs then just quits.. no msgs or anything. now if i set to join channel it will join announces but still that 60 sec pause or so.. Is there anyway to make the bot static? not worry about all the reconnects. drives me insane. I have never seen it identify to nickserv either.. so i just dont get those settings.

The bot was not designed to be "static". It was meant to connect, post and leave.
If you have issues with the bot not posing, you may add a check if the "StreamWriter status" is set to true. Also, it can easily be changed to be "static".

**Comment by costaht** [comment]
> The bot was not designed to be "static". It was meant to connect, post and leave. If you have issues with the bot not posing, you may add a check if the "StreamWriter status" is set to true. Also, it can easily be changed to be "static".

Obi-Wana, were you able to connect your bot to a server running TLS? Also, would you mind sharing how you made your bot keep the connection open? 
Thanks

**Comment by Obi-Wana** [comment]
> > The bot was not designed to be "static". It was meant to connect, post and leave. If you have issues with the bot not posing, you may add a check if the "StreamWriter status" is set to true. Also, it can easily be changed to be "static".
> 
> Obi-Wana, were you able to connect your bot to a server running TLS? Also, would you mind sharing how you made your bot keep the connection open? Thanks

I have only tested without TLS. And never tried to have the bot connected all the time.
